src/file_pipe_log/reader.rs

```diff
@@ -64,17 +65,54 @@ impl<F: FileSystem> LogItemBatchFileReader<F> {
     /// Returns the next [`LogItemBatch`] in current opened file. Returns
     /// `None` if there is no more data or no opened file.
     pub fn next(&mut self) -> Result<Option<LogItemBatch>> {
-        if self.valid_offset < self.size {
+        // @lucasliang.
+        // TODO: [Fulfilled in writing progress when DIO is open.]
+        // We should also consider that there might exists broken blocks when DIO
+        // is open, and the following reading strategy should tolerate reading broken
+        // blocks until it finds an accessible header of `LogBatch`.
+        while self.valid_offset < self.size {
+            let data_layout = self.file_context.as_ref().unwrap().format.data_layout();
             if self.valid_offset < LOG_BATCH_HEADER_LEN {
                 return Err(Error::Corruption(
                     "attempt to read file with broken header".to_owned(),
                 ));
             }
-            let (footer_offset, compression_type, len) = LogBatch::decode_header(&mut self.peek(
-                self.valid_offset,
-                LOG_BATCH_HEADER_LEN,
-                0,
-            )?)?;
+            let (footer_offset, compression_type, len) = {
+                match data_layout {
+                    DataLayout::NoAlignment | DataLayout::Alignment(0) => LogBatch::decode_header(
+                        &mut self.peek(self.valid_offset, LOG_BATCH_HEADER_LEN, 0)?,
+                    )?,
+                    DataLayout::Alignment(fs_block_size) => {
+                        // In DataLayout::Alignment mode, tail data in this block maybe
+                        // contain an valid header of the adjacent LogBatch or paddings
+                        // with `0`.
+                        let aligned_next_offset =
+                            round_up(self.valid_offset, fs_block_size as usize);
+                        let parser = LogBatch::decode_header(&mut self.peek(
```

Comment:
No, the point of using if_chain is to not match anymore. Two modes can share the decode_header (the hint is the same as well, in fact only in rare cases we need to access the end of block, we should optimize for common case where the decode_header succeeds).

Reply:
Roger that.